lahja/endpoint.py

```diff
@@ -248,25 +392,65 @@ def name(self) -> str:
         )
         self._server_running.set()
 
+    def receive_message(self, message: Broadcast) -> None:
+        self._receiving_queue.put_nowait((message.event, message.config))
+
     async def _accept_conn(self, reader: StreamReader, writer: StreamWriter) -> None:
-        remote = RemoteEndpoint(reader, writer)
-        coro = asyncio.ensure_future(self._handle_client(remote))
-        coro.add_done_callback(lambda fut: self._child_coros.remove(fut))
-        self._child_coros.add(coro)
+        conn = Connection(reader, writer)
+        remote = InboundConnection(conn, self.receive_message)
+        self._inbound_connections.add(remote)
 
-    async def _handle_client(self, remote: RemoteEndpoint) -> None:
-        while self._running:
-            try:
-                message = await remote.read_message()
-            except ReaderAtEof:
+        task = asyncio.ensure_future(remote.run())
+        task.add_done_callback(lambda _: self._inbound_connections.remove(remote))
+        task.add_done_callback(self._child_tasks.remove)
+        self._child_tasks.add(task)
+
+        # the Endpoint on the other end blocks until it receives this message
+        await remote.notify_subscriptions_updated(self.subscribed_events)
+
+    @property
+    def subscribed_events(self) -> Set[Type[BaseEvent]]:
+        "Return the set of events this Endpoint is currently listening for"
+        return set(self._handler.keys()) | set(self._queues.keys())
+
+    async def _notify_subscriptions_changed(self) -> None:
+        "Tell all inbound connections of our new subscriptions"
+        # make a copy so that the set doesn't change while we iterate over it
+        for inbound_connection in self._inbound_connections.copy():
+            await inbound_connection.notify_subscriptions_updated(self.subscribed_events)
+
+    async def wait_until_any_connection_subscribed_to(self,
+                                                      event: Type[BaseEvent]) -> None:
+        """
+        Blocks until an endpoint we've connected to wants our messages
+        """
+        if len(self._outbound_connections) == 0:
+            raise Exception('there are no outbound connections!')
+
+        for outbound in self._outbound_connections.values():
+            if event in outbound.subscribed_messages:
                 return
 
-            if isinstance(message, Broadcast):
-                await self._receiving_queue.put((message.event, message.config))
-            else:
-                self.logger.warning(
-                    f'[{self.ipc_path}] received unexpected message: {message}'
-                )
+        coros = [
+            outbound.wait_until_subscribed_to(event)
+            for outbound in self._outbound_connections.values()
+        ]
+        _, pending = await asyncio.wait(coros, return_when=asyncio.FIRST_COMPLETED)
+        (task.cancel() for task in pending)
+
+    async def wait_until_all_connections_subscribed_to(self,
```

Comment:
I like that suggestion.  Another thing that I think would be useful would be to be able to use a wildcard for the specific event types. Because, often, I think we do not even care about the specific event type because they all start listening at the same time. So something like this would do it:

```
endpoint.wait_until_acceptiong_events(('main', (AnyEvent)), ('blacklist-plugin', ((AnyEvent,)))
```

And I think, sometimes we really just care about this kind of thing:

```
endpoint.wait_until_anyone_listens_for_anything()
```

But let's move this into another PR :)